Updated Docs and Added Diagram (#46)
MetaMask/delegation-framework: documents/DeleGatorCore.md

Contains the logic needed for an ERC4337 SCA with delegation functionality. We provide two different "DeleGator implementations" for use: [MultiSigDeleGator](/documents/MultisigDeleGator.md) and [HybridDeleGator](/documents/HybridDeleGator.md). The distinction between the two is the signing mechanisms.

> The DeleGatorCore implements the UUPS proxy, which gets inherited in the implementations. There are two methods available to upgrade your account to a different implementation.

Reviewer: ```suggestion
The DeleGatorCore implements the UUPS proxy, which is inherited by the implementations. There are two methods available to upgrade your account to a different implementation.
```

Author: Thanks

> - `upgradeToAndCall` - This method upgrades the account to a different implementation defaulting to clearing the storage associated to the account's signature scheme.

Reviewer: ```suggestion
- `upgradeToAndCall` - This method upgrades the account to a different implementation, and by default will clear the storage associated with the account's signer configuration.
```

> - `upgradeToAndCallAndRetainStorage` - This method upgrades the account to a different implementation and retains the storage associated to the account's signature scheme.

Reviewer: ```suggestion
- `upgradeToAndCallAndRetainStorage` - This method upgrades the account to a different implementation and retains the storage associated with the account's signer configuration.
```

Author: thanks

# Signing a UserOperation

Contracts that extend the DeleGatorCore contract MUST use [EIP712](https://eips.ethereum.org/EIPS/eip-712) typed data signatures for User Operations to provide legibility when signing. The typed data to be signed is a [PackedUserOperation](https://github.com/eth-infinitism/account-abstraction/blob/releases/v0.7/contracts/interfaces/PackedUserOperation.sol).

## Rules

- A DeleGator Implementation MUST use namespaced storage for ALL variables if it is extending MetaMask's DeleGatorCore.
- DeleGator Implementations MUST inherit from DeleGatorCore as the "most base-like" (furthest right, [more info](https://docs.soliditylang.org/en/v0.8.23/contracts.html#multiple-inheritance-and-linearization))
- DeleGator Implementations MUST implement ERC-1271 Standard Signature Validation Method for Contracts
- DeleGator Implementations SHOULD return a non ERC-1271 magic value when encountering an invalid signature

> - DeleGator Implementations SHOULD implement a function to clear the storage related to the the signature scheme inside the function `_clearDeleGatorStorage`.

Reviewer: ```suggestion
- DeleGator Implementations SHOULD implement a function to clear the storage related to the account's signer configuration, inside the function `_clearDeleGatorStorage`.
```

If you don't like this suggestion, there was a duplicate "the" - `the the`

Author: thanks

## Rules

- A DeleGator Implementation MUST use namespaced storage for ALL variables if it is extending MetaMask's DeleGatorCore.
- DeleGator Implementations MUST inherit from DeleGatorCore as the "most base-like" (furthest right, [more info](https://docs.soliditylang.org/en/v0.8.23/contracts.html#multiple-inheritance-and-linearization))
- DeleGator Implementations MUST implement ERC-1271 Standard Signature Validation Method for Contracts
- DeleGator Implementations SHOULD return a non ERC-1271 magic value when encountering an invalid signature
- DeleGator Implementations SHOULD implement a function to clear the storage related to the account's signer configuration, inside the function `_clearDeleGatorStorage`.

> NOTE: If a DeleGator Implementation implements `reinitialize` it SHOULD gate the method to `onlySelf` to ensure no one can unexpectedly take over the DeleGator.

> > c

Reviewer: Should this 'c' be here?
```suggestion
```
